design/tls-backend-verification.md

```diff
@@ -0,0 +1,166 @@
+# TLS backend verification
+
+Status: Draft
+
+Contour 0.11 added the ability to communicate via TLS between Envoy and pods in the backend service.
+This proposal describes the facility for Envoy to verify the backend service's certificate.
+
+## Goals
+
+- Allow IngressRoute authors to assert that the backend service Envoy is communicating with is using a certificate signed by a certificate authority of their choice.
+- Allow IngressRoute authors to assert that the backend service Envoy is communicating with presents a subjectAltName of their choice.
+
+## Non Goals
+
+- Authentication via client side certificates is not in scope.
+- Support for k8s ingress documents.
+
+## Background
+
+Contour 0.11 added the `contour.heptio.com/upstream-protocol.tls` Service annotation which told Contour to use TLS when communicating with the members of the backend cluster, the Service's pods.
+However, this connection is not validated, Envoy will accept any certificate the client presents.
+
+This proposal seeks to improve the validation of TLS connections .
+
+## High-Level Design
+
+At a high level I propose the following:
+
+1. The CA to be used to validate the certificate presented by the backend should be packaged in a Secret.
+The secret will be placed in the namespace of the IngressRoute.
+
+2. IngressRoute's `spec.routes.services.service` entry will grow a new subkey, `validation`.
+
+3. The `validation` key will have a mandatory `caSecret` key which is the name of the secret where the ca to be validated is stored.
+Certificate Delegation is not in scope.
+
+4. The `validation` key will have an optional `subjectname` key which is expected to be present in the subjectAltName of the presented certificate.
+If `subjectname` is not present, any certificate with a valid chain to the supplied CA is considered valid.
+
+5. If `spec.routes.services[].validation` is present, `spec.routes.services[].{name,port}` must point to a service with a matching `contour.heptio.com/upstream-protocol.tls` Service annotation.
+If the Service annotation is not present or incorrect, the route is rejected with an appropriate status message.
+
+### Sample YAML
+
+```
+apiVersion: contour.heptio.com/v1beta1
+kind: IngressRoute
+metadata:
+  name: secure-backend
+spec:
+  virtualhost: 
+    fqdn: www.example.com  
+  routes:
+    - match: /
+      services:
+        - name: service
+          port: 8443
+          validation:
+            caSecret: my-certificate-authority
+            subjectname: backend.example.com 
+```
+
+## Detailed Design
+
+Implementation will consist of these steps:
+
+### CAs in Secrets
+
+The store of CA information is an opaque kubernetes secret.
+The secret will be stored in the same namespace as the corresponding IngressRoute.
+TLS certificate delegation is not in scope for this proposal.
+
+The secret object should contain one entry named `ca.key`, the constents will be the CA public key material.
```

Comment:
This should probably be named `ca.crt`. It does contain a public key, but it also contains other information required for certificate validation (the expiration date, a list of supported extensions, reference to a certificate revocation list, etc).